### PR TITLE
Documents for parallel analysis

### LINKED ISSUE #93: Documents for parallel analysis
**Is your feature request related to a problem? Please describe.**
Implement parallel analysis in MDAnalysis.

**Describe the solution you'd like**
a new page of information and tutorials.

**Related PR**
- Initial stage of parallel analysis by implementing https://github.com/MDAnalysis/mdanalysis/pull/2723

### PR DESCRIPTION
**Pull request by yuxuanzhuang**
Fixes #93 

The jupyter notebook can be read here:
https://yuxuanzhuang.github.io/UserGuide/examples/analysis/custom_parallel_analysis.html

### CODE REVIEW

**Comment by lilyminium** [comment]
> Last updated: Augus 2020
August?

> Any references should be used in the text and make use of citation syntax (not sure if/how this works in notebooks – ask @lilyminium ). At the moment, the References section can be removed because nothing is cited. For dask and multiprocessing just create links.

The [citation syntax](https://nbsphinx.readthedocs.io/en/0.4.1/markdown-cells.html#Citations) uses the html tag `data-cite="my_key_in_references_bib"`. I personally add it in brackets after the thing (e.g. `(#my_key_in_references_bib)` and run the `clean_example_notebooks.py` to replace it with the actual citation syntax and write my `References` section for me.

> and I think using MDAnalysisData is ok in the User Guide (@lilyminium ?). It has >4000 frames so parallelization makes more sense. 

Yes, although I note it in the "required packages" section :-)

**Comment by orbeckst** [comment]
@yuxuanzhuang it would be good to finish the ug entry before the end of GSoC as it is a major documentation piece, which was part of your roadmap.

**Comment by yuxuanzhuang** [comment]
Thanks for the reviewing!

Not sure it makes sense here, but I first describe how to do per-frame parallelism (which should normally be avoided), and then discuss a simple version of split-apply-combine scheme. 

Besides, should we cover (or point to somewhere else) about any pickling issues (e.g. what is not supported) and also what should be expected from returning `AtomGroup` (e.g. in https://github.com/MDAnalysis/pmda/issues/134#issuecomment-679110454).

@orbeckst  @richardjgowers @fiona-naughton @IAlibay any other suggestion?

**Comment by yuxuanzhuang** [comment]
Thanks for the revision! It seems not possible to attach (with markdown) an image to the jupyter notebook and display it properly on the website. I can only use `IPython.display` to achieve that (not in a pretty way).

**Review by orbeckst — CHANGES_REQUESTED** [review]
Sorry for the delay @yuxuanzhuang . Thank you for addressing my comments. I only have few remaining issues:

- I think you started using reST citations (following  @lilyminium 's example) so you can remove the _References_ section. 
- When you use the figure from the PMDA paper then you need to reference it and cite your source. ("Image from Ref [SFPG+19] used under XYZ license" – probably CC-BY but check).
- In the _See Also_ section, make PMDA a link to the PMDA docs.

**Comment by orbeckst** [comment]
@lilyminium do you have a better idea for including images than using `Image()`?

Does standard markdown not work, like
```markdown
![split/apply/combine scheme from [SFPG+19]](path/to/image.png)
```
?

**Comment by lilyminium** [comment]
[Yes](https://nbsphinx.readthedocs.io/en/0.7.1/markdown-cells.html#Images), there's an example in `examples/analysis/structure/dihedrals.ipynb` that points to `examples/images/dihedrals.png`:

```
<center>
<div style="width: 250px; text-align: center;">
    
![dihedral angles in proteins](../../images/dihedrals.png)

</div>
</center>
```

or you can also use the `` tag but I vaguely recall having problems centering images using HTML.

**Comment by yuxuanzhuang** [comment]
> [Yes](https://nbsphinx.readthedocs.io/en/0.7.1/markdown-cells.html#Images), there's an example in `examples/analysis/structure/dihedrals.ipynb` that points to `examples/images/dihedrals.png`:
> 
> ```
> <center>
> <div style="width: 250px; text-align: center;">
>     
> dihedral angles in proteins
> 
> </div>
> </center>
> ```
> 
> or you can also use the `` tag but I vaguely recall having problems centering images using HTML.

Thanks! I am not sure why it did not work before. Now it looks good :)

**Review by orbeckst — APPROVED** [review]
Thanks for the updates, looking good to me.

(You still have the _References_ section that is technically not needed any more for the online docs version and you could remove it. I don't know if @lilyminium has particular guidelines on if references should be doubled up so that the notebook can stand on its own, too.)

**Comment by yuxuanzhuang** [comment]
> Thanks for the updates, looking good to me.
> 
> (You still have the _References_ section that is technically not needed any more for the online docs version and you could remove it. I don't know if @lilyminium has particular guidelines on if references should be doubled up so that the notebook can stand on its own, too.)

I did remove the References section. If I understand it correctly after I ran the `clean_example_notebooks.py`, it reappeared automatically.
The https://yuxuanzhuang.github.io/UserGuide/examples/analysis/custom_parallel_analysis.html should be its final version.

**Comment by orbeckst** [comment]
Oh, ok. Thank you for explaining. I have no more comments. 

As far as I am concerned, @lilyminium can merge.

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 1]
```diff
@@ -0,0 +1,986 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def radgyr_per_frame(frame_index, atomgroup, masses, total_mass=None):\n",
+    "    \n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:214]:
```suggestion
    "    # index the trajectory to set it to the frame_index frame\n",
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 2]
```diff
@@ -0,0 +1,986 @@
...
+    "    return np.sqrt(rog_sq)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parallelization with multiprocessing\n",
+    "\n",
+    "The native parallel module in python is called [multiprocessing](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing), it contains some useful tools to build a pool of working cores, map the function into different workers, and gather and order the results from all the workers.\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:244]:
```suggestion
    "The native parallelisation module in Python is called [multiprocessing](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing). It contains useful tools to build a pool of working cores, map the function into different workers, and gather and order the results from all the workers.\n",
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 3]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "Below we use `Pool` from `multiprocessing` as a context manager, we can define how many cores (or workers) we want to use by `Pool(n_jobs)`. Then we can map our analysis method `radgyr_per_frame` with the parameters needed with `partial`. Note we do not include `frame_index` inside partial, because `frame_index` is the variable that is iterated to create a list of jobs.\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:246]:
```suggestion
    "Below we use `Pool` from `multiprocessing` as a context manager. We can define how many cores (or workers) we want to use with `Pool(n_jobs)`. We use `functools.partial` to create a new method by supplying every argument needed for `radgyr_per_frame` *except* `frame_index`. We can do this because the `atomgroup`, `masses` etc. will not change when we iterate the function over each frame, but the `frame_index` will. We create a list of jobs where we use the `worker_pool` to map each `frame_index` to each job.\n",
```

Just clarifying that `functools.partial` is quite separate from using the `Pool` context manager.

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 4]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "The downside of `multiprocessing` is that it is mostly focused on single-machine multicore parallelism (without extra manager). It is hard to operate on multimachine conditions (we use `dask` instead) "
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:248]: Not sure this fits here, after the description of the function but before the actual code. It may work better as an intro to the dask section?

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 5]
```diff
@@ -0,0 +1,986 @@
...
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with Pool(n_jobs) as worker_pool:\n",
+    "    result = worker_pool.map(partial(radgyr_per_frame, \n",
+    "                                     atomgroup=protein,\n",
+    "                                     masses=protein.masses,\n",
+    "                                     total_mass=np.sum(protein.masses)),\n",
+    "                             range(u.trajectory.n_frames))"
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:273]: Can you make the partial function a different thing outside the context manager to separate these concepts out more, e.g.

```python
run_per_frame = partial(radgyr_per_frame, atomgroup=....)
frame_values = np.arange(u.trajectory.n_frames)
with Pool(n_jobs) as worker_pool:
    result = worker_pool.map(run_per_frame, frame_values)
```
?

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 6]
```diff
@@ -0,0 +1,986 @@
...
+    "plt.xlabel('Frame');"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parallelization with dask\n",
+    "\n",
+    "[Dask](https://docs.dask.org/en/latest/) is a flexible library for parallel computing in Python. It provides advanced parallelism for analytics that has been integrated or utilized in many scientific softwares. It can be scaled from one single computer to a cluster of computers inside a HPC center.\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:319]:
```suggestion
    "[Dask](https://docs.dask.org/en/latest/) is a flexible library for parallel computing in Python. It provides advanced parallelism for analytics and has been integrated or utilized in many scientific softwares. It can be scaled from one single computer to a cluster of computers inside a HPC center.\n",
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 7]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "It has a dynamic task scheduling system that can be used either in synchronous (single-core), threading, multiprocessing or distributed cluster. The wrapping function in dask, `dask.delayed`, mimics for loops and wraps custom normal python code into a dask graph. Then it can be easily run in parallel, and visualized by `dask.visualize()` to examine if the task is well distributed. To be more precise, the code inside`dask.delayed` will not run immediately but be pushed into a job queue waiting for submission (`compute()` or `persist()`). A more complex dependency can also be built among different jobs, which might be useful for a complex analysis.\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:321]:
```suggestion
    "Dask has a dynamic task scheduling system with synchronous (single-threaded), threaded, multiprocessing and distributed schedulers. The wrapping function in dask, `dask.delayed`, mimics for loops and wraps Python code into a Dask graph. This code can then be easily run in parallel, and visualized with `dask.visualize()` to examine if the task is well distributed. The code inside `dask.delayed` is not run immediately on execution, but pushed into a job queue waiting for submission. These queues can be resolved or executed with commands like `compute()` or `persist()`. A more complex dependency can also be built among different jobs, which might be useful for a complex analysis.\n",
```

I find this very short explanation a bit confusing, especially the bits at the end ("A more complex dependency..." expecially. Could you go into more detail or just link to Dask resources explaining these concepts?

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 8]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "Below are two simple examples to use Dask to achieve the same task as `multiprocessing` does.\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:323]: Here would be a good place to move that earlier part in the multiprocessing that talks about its disadvantages.

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 9]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "The API of Dask is similar to `multiprocessing`. It also creates a pool of workers for your single machine with the given resources.\n",
+    "\n",
+    "Note: The threaded scheduler in dask (similar to `threading` in python) should not be used as it will mess up with the state (timestep) of the trajectory."
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:327]:
```suggestion
    "Note: The threaded scheduler in Dask (similar to `threading` in Python) should not be used as it will mess up with the state (timestep) of the trajectory."
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 10]
```diff
@@ -0,0 +1,986 @@
...
+    "plt.legend()\n",
+    "plt.ylabel('Radius of gyration (Å)')\n",
+    "plt.xlabel('Frame');"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can also use the old `radgyr` function because `dask` is more flexible on the input args.\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:487]:
```suggestion
    "We can also use the old `radgyr` function because `dask` is more flexible on the input arguments.\n",
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 11]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "Note: `protein` will change its state during the trajectory iteration, so the processes are always aware which timestep the trajectory is in and change the `protein` (e.g. positions) accordingly."
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:489]: Can you rearrange this to say `coordinates`, as it's ambiguous how protein is changed?

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 12]
```diff
@@ -0,0 +1,986 @@
...
+    "plt.legend()\n",
+    "plt.ylabel('Radius of gyration (Å)')\n",
+    "plt.xlabel('Frame');"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can also use Dask dashboard (with dask.distributed.Client) to examine how jobs are distributed along all the workers. So each green bar below represents a single-frame `radgyr`."
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:554]:
```suggestion
    "We can also use Dask dashboard (with dask.distributed.Client) to examine how jobs are distributed along all the workers. Each green bar below represents one job, i.e. running `radgyr` on one frame of the trajectory."
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 13]
```diff
@@ -0,0 +1,986 @@
...
+    "</center>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Parallelization in a split-apply-combine fashion\n",
+    "\n",
+    "The aforementioned per-frame approach should normally be **avoided** because in **each** task, all the attributes (`AtomGroup`, `Universe`, and etc) need to be pickled. This might even takes more time than your lightweight analysis! Besides, e.g. in dask, a huge amount of time is needed overhead to build a comprehensive dask graph with thousands of tasks.\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:576]:
```suggestion
    "The aforementioned per-frame approach should normally be **avoided** because in **each** task, all the attributes (`AtomGroup`, `Universe`, and etc) need to be pickled. This pickling may take even more time than your lightweight analysis! Besides, in Dask, a significant amount of overhead time is needed to build a comprehensive Dask graph with thousands of tasks.\n",
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 14]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "So a simple split-apply-combine scheme for parallel trajectory analysis should normally be applied (See the figure below)--the trajectory is split into blocks, analysis is performed separately and in parallel on each block (\"apply\"), then results from each block are gathered and combined. \n"
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:578]:
```suggestion
    "Therefore, we should normally use a split-apply-combine scheme for parallel trajectory analysis. Here, the trajectory is **split** into blocks, analysis is performed separately and in parallel on each block (\"apply\"), and then results from each block are gathered and **combined**. \n"
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 15]
```diff
@@ -0,0 +1,986 @@
...
+    "### Block analysis function"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@dask.delayed\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:615]: Can you just briefly explain that dask.delayed as a decorator here is the same as `dask.delayed(analyze_block)` or something like that?

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 16]
```diff
@@ -0,0 +1,986 @@
...
+   "metadata": {},
+   "source": [
+    "### Split the trajectory"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This is a very simple way to split the trajectory, sometime it is not really balanced (e.g. the last block)."
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:635]: Can you briefly explain that the code below is splitting it into equal blocks?

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 17]
```diff
@@ -0,0 +1,986 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_frames = u.trajectory.n_frames\n",
+    "n_blocks = n_jobs   #  it can be any realistic value (0<n_blocks<=n_jobs)\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:645]:
```suggestion
    "n_blocks = n_jobs   #  it can be any realistic value (0 < n_blocks <= n_jobs)\n",
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 18]
```diff
@@ -0,0 +1,986 @@
...
+    "\n",
+    "frame_per_block = n_frames // n_blocks\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:647]:
```suggestion
    "n_frames_per_block = n_frames // n_blocks\n",
```

This is longer but clearer. Alternatively you could comment the meaning

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 19]
```diff
@@ -0,0 +1,986 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "jobs = []\n",
+    "for bs in blocks:\n",
+    "    jobs.append(analyze_block(bs, radgyr, protein, protein.masses, total_mass=np.sum(protein.masses)))\n",
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:698]: Can you add a newline so the code follows pep8?

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 20]
```diff
@@ -0,0 +1,986 @@
...
+    "jobs = dask.delayed(jobs)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "By `visualize()` we can confirm that the trajectory is split into a few blocks instead of ~4000 jobs."
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:706]:
```suggestion
    "Using `visualize()`, we can see that the trajectory is split into a few blocks instead of ~4000 jobs."
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 21]
```diff
@@ -0,0 +1,980 @@
...
...i/S/6rjOcevNeccjkeFXBdWgNJUetUD/hV18RaJpEIAABCAAAQhAAAIQgAAE9qYA8oWf1X9hzvFQZPi1n2ocNV8fYlrk1KlO/b8UVMMQc2He0bHAdSFfyO/NawqthgAEIAABCEAAAhCAAAQgcBAFfnq+MGO/JzYcOoD9F5AvvPd6++WXXxoaGvJ55AvvJcIHEIAABCAAAQhAAAIQgAAEdpvA1+cLfeRJibH2i5+PQL6A/gtvXhTIF94UwXsIQAACEIAABCAAAQhAAAK7XgD5Ap6PwPiOu/4yRQMhAAEIQAACEIAABCAAAQjseoFdmC8MmM6I9P+378d3xPMR77040H/hvTT4AAIQgAAEIAABCEAAAhCAwG4VQL6w2/ovKOlmseHQdxrfcdDU9Gj7vzF/xG69HNEuCEAAAhCAAAQgAAEIQAACe1YA+cJPzxdGzFc79f9bmT8C+UIR/Rf27P0EDYcABCAAAQhAAAIQgAAEDq5AvpDlYoZFTtxjPKLjOucdHZO2WxKidsz8h5a9JCXqJq1/j1tu9hiPLDmldHg1m0/zRT6eDW0HJqftd7qJQyr6vNRYt+SUGHZmiJ3ZaftdqfHwklO86VXPOR6JiZph9rKauViu8/qI+UqP8cgTT78rTsaygZe+xzP2exKibsZ+Z9x6o9vwu5q5qGHbuvS/qejzEkPtklPChNd5ns8VMmx4fcZ+T0rULXDdT72D845HEqK0X0tkK89nrdEtHdcprKFDK0xobdZ+v/xW6oi+rD7BuUKG2Jkrf1q74ux1xgzVn1Yv5/mcOfJ01v5AQtQuOcWWyEb1p28sF/icJbIxz5UOedEpZkJrQoECn7dENuYcD4X1lshmMhfWBybnHA+7DL+Nmq8/tt4Q6f81SDd16f9viGkdZi+LiZpNr4YM6lZcsm7D7yPmqxr2Yne5U4PeP/3MNyKMpvncN/rcNzLneNRt+H3C+te49WaX4VcNc1FrvtRZqvBcl+G3Ba573T0w73gkMvyqZi4Om6906v9XSTc92v5/A6azHfr/EZ5G0TlEq2ZVNBJ946A+9+1/fe4G7yyPfOGdLFgJAQhAAAIQgAAEIAABCEBglwvkC1lz+KmCajAFl4Mpbjsw2W9qtEWfG3amFVSDPfrCHHkqp+rp0EoyFxGOpcAXwhnPlk8rp+pf+Mf6yFNUcDGeDWXyCUNgRk7VO+NE
```
[the remark this answers is not in the document]
Comment (lilyminium) [reply doc/source/examples/analysis/custom_parallel_analysis.ipynb:788]: represents typo still here

File: doc/source/reading_and_writing.rst [hunk 22]
```diff
@@ -191,21 +191,46 @@ You can pass keyword arguments to some format writers. For example, the :ref:`LA
...
+    from MDAnalysis.tests.datafiles import PSF, DCD
+    u = mda.Universe(PSF, DCD)
+    g = u.atoms
+    g_pickled = pickle.loads(pickle.dumps(g))
+    print("g_pickled.universe is u: ", u is g_pickled.universe)
+    g_pickled, u_pickled = pickle.loads(pickle.dumps((g, u)))
+    print("g_pickled.universe is u_pickled: ",
+           u_pickled is g_pickled.universe)
+
+If multiple :class:`MDAnalysis.core.groups.AtomGroup` are bound to the same
```
Comment (lilyminium) [inline doc/source/reading_and_writing.rst:221]:
```suggestion
If multiple :class:`MDAnalysis.core.groups.AtomGroup`\ s are bound to the same
```

File: doc/source/reading_and_writing.rst [hunk 23]
```diff
@@ -191,21 +191,46 @@ You can pass keyword arguments to some format writers. For example, the :ref:`LA
...
+:class:`MDAnalysis.core.universe.Universe`, they will bound to the same one
```
Comment (lilyminium) [inline doc/source/reading_and_writing.rst:222]:
```suggestion
:class:`MDAnalysis.core.universe.Universe`, they will also be bound to the same one
```

File: doc/source/reading_and_writing.rst [hunk 24]
```diff
@@ -191,21 +191,46 @@ You can pass keyword arguments to some format writers. For example, the :ref:`LA
...
 
     import pickle
     from MDAnalysis.tests.datafiles import PSF, DCD
     psf = mda.Universe(PSF, DCD)
-    pickle.loads(pickle.dumps(psf.trajectory))
+    pickle.loads(pickle.dumps(psf))
 
-While *trajectories* from PDB files cannot be pickled, trajectories where only the topology information comes from a PDB file *can*. For example, the universe below loads the trajectory information from a :ref:`TRR <TRR-format>` file.
+As for :class:`MDAnalysis.core.groups.AtomGroup`, during serialization, it will be pickled with its bound
+:class:`MDAnalysis.core.universe.Universe` which means after unpickling,
```
Comment (lilyminium) [inline doc/source/reading_and_writing.rst:204]:
```suggestion
:class:`MDAnalysis.core.universe.Universe`. This means that after unpickling,
```

File: doc/source/reading_and_writing.rst [hunk 25]
```diff
@@ -191,21 +191,46 @@ You can pass keyword arguments to some format writers. For example, the :ref:`LA
...
+a new :class:`MDAnalysis.core.universe.Universe` will be created and
+be attached by the new :class:`MDAnalysis.core.groups.AtomGroup`. If the Universe is serialized
```
Comment (lilyminium) [inline doc/source/reading_and_writing.rst:206]:
```suggestion
be attached to the new :class:`MDAnalysis.core.groups.AtomGroup`. If the Universe is serialized
```

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb [hunk 26]
```diff
@@ -0,0 +1,986 @@
...
...dt0JuIoN4cy0+IxUmde+65ys7OVp06dXTJJZfo3XfftR4JPtm/f7/uuOMODRkyRA8++KDefPNNfrS8j2666SZlZGRo+vTpSklJ0dq1a61Hgk++/vprpaSkaMaMGcrIyNAvfvEL65FiRoUKFTR27Fj9+te/1i233KK77rpLBw4csB4LPnn77bfVsWNH1a9fX9nZ2WrRooX1SDGD3oxd9KYtejN20Zt26M3YRm/aoTdjF71pi96MXfSmHXozttGbdujN2EVv2qI3Yxe9aYfejG2R1pvcEItTqlOnjmbPnq0777xT1157rYYOHaqDBw9ajwUPfffdd+rSpYvefPNNjRs3Tn/84x8VCoWsx4o53bp10xdffKHy5cvrggsu0MSJE61HgscyMjLUvn17hUIhzZ8/X127drUeKeaEQiE98cQTmjBhgt58802lpqZq/fr11mPBQ/n5+Xr44Yc1YMAAXXfddZo5c6Zq165tPVbMoTdjD70ZDPRm7KE37dGbsYfeDAZ6M/bQm8FAb8YeetMevRl76M1goDdjD70ZDPRm7KE37dGbsSdie9P6Z9QicowePdpVqFDBXXHFFW7Tpk3W48ADM2bMcHXq1HHnnXee+/LLL63HgTv84/5vuOEGFxcX5/7whz+4Q4cOWY+EMDt06JD7/e9/7+Li4twvfvEL/lmPgFi+fLlr2bKlq1u3rps5c6b1OPDA999/77p06eIqVqzo/v3vf1uPg/+hN6MfvRk89Gb0ozeDid6MfvRmMNGb0Y/eDB56M/rRm8FEb0Y/ejOY6M3oR28GD70Z/ejNYKI3o18E9+ZYbohFiSxcuNA1adLE1alTx02ePNl6HITJ/v373QMPPODi4uJc//793c6dO61HwnGGDx/uEhMT3eWXX+7Wr19vPQ7CZN26dS4tLc1VqFDB/fOf/7QeB8fJzc1111xzjYuLi3MPP/ywO3DggPVICJOJEye62rVru2bNmrlFixZZj4Pj0JvRid4MPnozOtGbwUZvRi96M9jozehEbwYfvRmd6M1gozejF70ZbPRmdKI3g4/ejE70ZrDRm9ErwnuTG2JRcjt37nS33Xabk+QGDRrkdu/ebT0SyuDLL790F110katYsaL761//aj0OTmHZsmWubdu2rkqVKu7111+3HgdlNH78eFe9enXXqlWrSAyImPLaa6+5ypUru3bt2rmVK1
```
Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:812]:
```suggestion
    "Here we use [joblib](https://joblib.readthedocs.io/en/latest/). It is implemented on `multiprocessing` and provides lightweight pipelining in Python. Compared to `multiprocessing`, it has a simple API and convenient persistence of cached results."
```

Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:810]:
```suggestion
    "You may want to perform multiple analyses (or analyze multiple trajectories). In this case, you can use some high-level parallelism, i.e. running all the serial analyses in parallel.\n",
```

Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:843]: Why is this here? Can you use a Markdown cell to link to and explain this?

Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:879]: This section could benefit from some general explanation between each cell. e.g. the necessity of `run_analysis`, and what `delayed(run_analysis)(analysis)` is doing.

Comment (lilyminium) [inline doc/source/examples/analysis/custom_parallel_analysis.ipynb:918]:
```suggestion
    "The parallel version of MDAnalysis is still under development. For existing solutions and some implementations of parallel analysis, go to [PMDA](https://www.mdanalysis.org/pmda/). PMDA (<a data-cite=\"fan_pmda_2019\" href=\"https://doi.org/10.25080/Majora-7ddc1dd1-013\">Shujie *et al.*, 2019</a>) applies the aforementioned split-apply-combine scheme with Dask. In the future, it may provide a framework that consolidates all the parallelisation schemes described in this tutorial."
```

**Review by lilyminium — CHANGES_REQUESTED** [review]
Thanks @yuxuanzhuang for adding this, it's a great addition to MDAnalysis and the tutorial is super thorough! I just have some notes on how you could improve grammar and clarity.

File: doc/source/examples/analysis/custom_parallel_analysis.ipynb (hunk 26 above)
Comment (lilyminium) [reply doc/source/examples/analysis/custom_parallel_analysis.ipynb:812]: Uh, I don't know why GitHub put this and the below comment on this images section. This suggestion pertains to the `joblib` part.

Comment (lilyminium) [reply doc/source/examples/analysis/custom_parallel_analysis.ipynb:810]: This is supposed to be a suggestion for the part that starts "In real life"...

Comment (lilyminium) [reply doc/source/examples/analysis/custom_parallel_analysis.ipynb:843]: Um, this is supposed to be about the comment link to the user guide. Sorry that all my suggestions appear to have been put on the same image output; I can't seem to change it.

Comment (lilyminium) [reply doc/source/examples/analysis/custom_parallel_analysis.ipynb:879]: This is about the general `joblib` section

Comment (lilyminium) [reply doc/source/examples/analysis/custom_parallel_analysis.ipynb:918]: This is about the last parallel section part.

**Comment by lilyminium** [comment]
Thank you @yuxuanzhuang, it looks great! Just a couple formatting issues and then it's good to go:

* `In this case, you can use some high-level parallelism, i.e. running all the serial analyses in parallel.:nbsphinx-math:n”,` seems to be broken? I can see the nbsphinx-math tag that I'm assuming should be part of the theme instead of something you manually put in.

* ` PMDA ([“fan_pmda_2019”]) applies the aforementioned split-apply-combine scheme with Dask. In the future, it may provide a framework that consolidates all the parallelisation schemes described in this tutorial.”` Similarly not sure what's going on here -- maybe the Python script to add references is choking up? If so, try `(fan_pmda_2019)` as iirc I wrote it for that format.

**Comment by yuxuanzhuang** [comment]
Thanks for the review! Seems I copied extra backslashes from suggestions.

**Review by lilyminium — APPROVED** [review]

**Comment by lilyminium** [comment]
Thanks @yuxuanzhuang! Could you just add yourself to AUTHORS and we can merge? This is great :-)

**Comment by yuxuanzhuang** [comment]
Thanks for all the reviews!

**Comment by lilyminium** [comment]
Thank you @yuxuanzhuang!
